pkg/slayers/scion.go

```diff
@@ -598,48 +604,103 @@ func scionNextLayerTypeL4(t L4ProtocolType) gopacket.LayerType {
 }
 
 // DstAddr parses the destination address into a addr.Host.
-func (s *SCION) DstAddr() (addr.Host, error) {
+// @ requires acc(&s.DstAddrType, R20) && acc(&s.RawDstAddr, R20)
+// @ requires s.DstAddrType == T4Svc ==> len(s.RawDstAddr) >= HostLenSVC
+// @ requires acc(sl.Bytes(s.RawDstAddr, 0, len(s.RawDstAddr)), R15)
+// @ ensures  acc(&s.DstAddrType, R20) && acc(&s.RawDstAddr, R20)
+// @ ensures  err == nil ==> IsHostTypeIP(res)  || IsHostTypeSVC(res)
+// @ ensures  err != nil ==>
+// @ 	acc(sl.Bytes(s.RawDstAddr, 0, len(s.RawDstAddr)), R15)
+// @ ensures  err != nil ==> err.ErrorMem()
+// @ decreases
+func (s *SCION) DstAddr() (res addr.Host, err error) {
 	return ParseAddr(s.DstAddrType, s.RawDstAddr)
 }
 
 // SrcAddr parses the source address into a addr.Host.
-func (s *SCION) SrcAddr() (addr.Host, error) {
+// @ requires  acc(&s.SrcAddrType, R20) && acc(&s.RawSrcAddr, R20)
+// @ requires  s.SrcAddrType == T4Svc ==> len(s.RawSrcAddr) >= HostLenSVC
+// @ requires  acc(sl.Bytes(s.RawSrcAddr, 0, len(s.RawSrcAddr)), R15)
+// @ ensures   acc(&s.SrcAddrType, R20) && acc(&s.RawSrcAddr, R20)
+// @ ensures  err == nil ==> IsHostTypeIP(res)  || IsHostTypeSVC(res)
+// @ ensures   err != nil ==>
+// @ 	acc(sl.Bytes(s.RawSrcAddr, 0, len(s.RawSrcAddr)), R15)
+// @ ensures   err != nil ==> err.ErrorMem()
+// @ decreases
+func (s *SCION) SrcAddr() (res addr.Host, err error) {
 	return ParseAddr(s.SrcAddrType, s.RawSrcAddr)
 }
 
 // SetDstAddr sets the destination address and updates the DstAddrType field accordingly.
-func (s *SCION) SetDstAddr(dst addr.Host) error {
+// @ requires  acc(&s.RawDstAddr)
+// @ requires  acc(&s.DstAddrType)
+// @ ensures   IsHostTypeSVC(dst) ==> res == nil
+// @ ensures   acc(&s.RawDstAddr) && acc(&s.DstAddrType)
```

Comment:
This spec can be compressed
```suggestion
// @ preserves acc(&s.RawDstAddr) && acc(&s.DstAddrType)
// @ ensures   IsHostTypeSVC(dst) ==> res == nil
```

Reply:
Done. 

sIs `preserves X` really just syntactic sugar for `requires X` and `ensures X`? If so I will also adapt the spec for `SrcAddr` and `SrcAddr`.

Reply:
yes, but we should guarantee that all our clauses are structured always in the order `requires`, `preserves`, `ensures`. We should not mix the orders

Reply:
I see. I have additionally adapted the spec in `ParseAddr`. 
`DstAddr` and `SrcAddr` I left as is.